src/number_theory/pell.lean

```diff
@@ -241,6 +144,217 @@ lemma x_neg (a : solution₁ d) : (-a).x = -a.x := rfl
+begin
+  intro hy,
+  have prop := a.prop,
+  rw [hy, sq (0 : ℤ), zero_mul, mul_zero, sub_zero] at prop,
+  exact lt_irrefl _ (((one_lt_sq_iff $ zero_le_one.trans ha.le).mpr ha).trans_eq prop),
+end
+
+/-- A solution with `x = 1` is trivial. -/
+lemma eq_one_of_x_eq_one (h₀ : 0 < d) {a : solution₁ d} (ha : a.x = 1) : a = 1 :=
+begin
+  have prop := a.prop_y,
+  rw [ha, one_pow, sub_self, mul_eq_zero, or_iff_right h₀.ne', sq_eq_zero_iff] at prop,
+  exact ext ha prop,
+end
+
+/-- A solution is `1` or `-1` if and only if `y = 0`. -/
+lemma eq_one_or_neg_one_iff_y_eq_zero {a : solution₁ d} : a = 1 ∨ a = -1 ↔ a.y = 0 :=
+begin
+  refine ⟨λ H, H.elim (λ h, by simp [h]) (λ h, by simp [h]), λ H, _⟩,
+  have prop := a.prop,
+  rw [H, sq (0 : ℤ), mul_zero, mul_zero, sub_zero, sq_eq_one_iff] at prop,
+  exact prop.imp (λ h, ext h H) (λ h, ext h H),
+end
+
+/-- The set of solutions with `x > 0` is closed under multiplication. -/
+lemma x_mul_pos_of_x_pos (h₀ : 0 < d) {a b : solution₁ d} (ha : 0 < a.x) (hb : 0 < b.x) :
+  0 < (a * b).x :=
+begin
+  simp only [x_mul],
+  refine neg_lt_iff_pos_add'.mp (abs_lt.mp _).1,
+  rw [← abs_of_pos ha, ← abs_of_pos hb, ← abs_mul, ← sq_lt_sq, mul_pow a.x, a.prop_x, b.prop_x,
+      ← sub_pos],
+  ring_nf,
+  positivity,
+end
+
+/-- The set of solutions with `x` and `y` positive is closed under multiplication. -/
+lemma x_mul_pos_and_y_mul_pos_of_x_pos_of_y_pos (h₀ : 0 < d) {a b : solution₁ d} (hax : 0 < a.x)
+  (hay : 0 < a.y) (hbx : 0 < b.x) (hby : 0 < b.y) :
+  0 < (a * b).x ∧ 0 < (a * b).y :=
+begin
+  simp only [x_mul, y_mul],
+  split; positivity,
+end
+
+/-- If `(x, y)` is a solution with `x` and `y` positive, then all powers of it have positive `x`,
+and the sign of `y` agrees with that of the exponent. -/
+lemma x_pow_pos_of_x_pos_of_y_pos (h₀ : 0 < d) {a : solution₁ d} (hax : 0 < a.x) (hay : 0 < a.y)
+  (n : ℤ) :
+  0 < (a ^ n).x ∧ (a ^ n).y.sign = n.sign :=
+begin
+  have H : ∀ m : ℤ, 0 < m → 0 < (a ^ m).x ∧ 0 < (a ^ m).y,
+  { change ∀ m : ℤ, 1 ≤ m → _,
+    refine λ m, int.le_induction (by simp only [hax, hay, zpow_one, and_self]) (λ m hm ih, _) m,
+    rw zpow_add_one,
+    exact x_mul_pos_and_y_mul_pos_of_x_pos_of_y_pos h₀ ih.1 ih.2 hax hay, },
+  rcases lt_trichotomy 0 n with hpos | rfl | hneg,
+  { rw [(int.sign_eq_one_iff_pos n).mpr hpos, int.sign_eq_one_iff_pos],
+    exact H n hpos, },
+  { rw [int.sign_zero, int.sign_eq_zero_iff_zero, zpow_zero],
+    simp only [x_one, zero_lt_one, y_one, eq_self_iff_true, and_self], },
+  { rw [(int.sign_eq_neg_one_iff_neg n).mpr hneg, int.sign_eq_neg_one_iff_neg, ← neg_neg n,
+        zpow_neg, x_inv, y_inv, neg_lt, neg_zero],
+    exact H (-n) (lt_neg.mp hneg), }
+end
+
+/-- If `a` is any solution, then one of `a`, `a⁻¹`, `-a`, `-a⁻¹` has
+positive `x` and nonnegative `y`. -/
+lemma exists_pos_variant (h₀ : 0 < d) (a : solution₁ d) :
+  ∃ b : solution₁ d, 0 < b.x ∧ 0 ≤ b.y ∧ (a = b ∨ a = b⁻¹ ∨ a = -b ∨ a = -b⁻¹) :=
```

Comment:
Optionally:
```suggestion
  ∃ b : solution₁ d, 0 < b.x ∧ 0 ≤ b.y ∧ a ∈ {b, b⁻¹, -b, -b⁻¹} :=
```
which is `defeq` but perhaps easier to read. Optimistically no change to the proof needed.

Reply:
To make Lean accept it, it seems that I need to write `({b, b⁻¹, -b, -b⁻¹} : set (solution₁ d))`.
Is it still worth it then? (The proof seems indeed to be OK, though.)

Reply:
Up to you. It will look nicer in the web docs but worse in the source code.

Reply:
I'll go with the set for now. If it gives me problems later (which I don't expect), I can change it back.